File: backend/internal/authjwt/dispatcher.go
package authjwt

import (
	"fmt"

	"github.com/golang-jwt/jwt/v5"
)

// LocalHMACIssuer/LocalRSAIssuer はbffが自分で発行するローカルJWTの`iss`固定値
// (CONTRACT.md セクション16.4)
// bff側もこれと同じ文字列をissとして設定する
const (
	LocalHMACIssuer = "bff-gin-local-hmac"
	LocalRSAIssuer  = "bff-gin-local-rsa"
)

// IsLocalIssuer はclaims.Issuerがローカル発行(HMAC/RSAいずれか)かどうかを返す
// v1/task.go・grpcserver/task_service.goのresolveUserIDが、Keycloak発行トークン
// (sub=keycloak_sub)とローカル発行トークン(sub=内部user_idそのもの)を
// 区別するために使う(CONTRACT.mdセクション16.5)
func IsLocalIssuer(iss string) bool {
	return iss == LocalHMACIssuer || iss == LocalRSAIssuer
}

// TokenVerifier はJWT検証の最小インターフェース
// *Verifier(JWKS/RSA)・*HMACVerifier・*Dispatcher のいずれもこれを満たす
type TokenVerifier interface {
	Verify(tokenString string) (*Claims, error)
}

// Dispatcher はJWTの`iss`クレームで検証方式(Keycloak/ローカルHMAC/ローカルRSA)を
// 振り分ける(CONTRACT.md セクション16.5)
//
// 【設計上の要点】ParseUnverifiedで覗いたissは署名検証前の値であり、誰でも自由に書き換えられる
// ここではその値を「どのverifierに委譲するか」を決める振り分け
// キーとしてのみ使い、実際の信頼(本当にそのissuerが署名したか)は委譲先の verifier.Verifyが行う署名検証に委ねる
// この2段階構造により、未知の iss を詐称しても対応するverifierが存在しないため即エラーになり、
// 既知のissを詐称しても対応するverifierの署名検証で弾かれる
type Dispatcher struct {
	byIssuer map[string]TokenVerifier
}

func NewDispatcher() *Dispatcher {
	return &Dispatcher{byIssuer: make(map[string]TokenVerifier)}
}

// Register はissuer文字列とverifierの対応を登録する
// メソッドチェーンできるよう自分自身を返す
func (d *Dispatcher) Register(issuer string, verifier TokenVerifier) *Dispatcher {
	d.byIssuer[issuer] = verifier
	return d
}

func (d *Dispatcher) Verify(tokenString string) (*Claims, error) {
	var peek jwt.RegisteredClaims
	if _, _, err := jwt.NewParser().ParseUnverified(tokenString, &peek); err != nil {
		return nil, fmt.Errorf("JWTのパースに失敗(iss確認前): %w", err)
	}
	verifier, ok := d.byIssuer[peek.Issuer]
	if !ok {
		return nil, fmt.Errorf("不明なissuer: %q", peek.Issuer)
	}
	return verifier.Verify(tokenString)
}

File: backend-elixir/lib/backend_elixir/auth/dispatcher.ex
defmodule BackendElixir.Auth.Dispatcher do
  @moduledoc """
  JWTの`iss`クレームで検証方式(Keycloak/ローカルHMAC/ローカルRSA)を振り分ける
  (backend(Go)のauthjwt.Dispatcher・backend-rust/backend-c/backend-cpp/backend-java/
  backend-kotlin/backend-pythonと同じ2段構造)。issの詐称は委譲先の署名検証で弾かれる:
  「振り分けのために覗く」ことと「検証を信頼する」ことは別、という2段階構造を維持する。

  検証方式(HmacVerifier/JwksVerifier)は`{module, arg}`のタプルとして登録する。verify/2は
  登録されたmoduleの`verify/2`関数を呼ぶ(Elixirにはinterfaceが無いため、共通の関数名を
  持つモジュールを値として保持するダックタイピング的な設計にする)
  """

  alias BackendElixir.Auth.Claims

  @local_hmac_issuer "bff-gin-local-hmac"
  @local_rsa_issuer "bff-gin-local-rsa"

  def local_hmac_issuer, do: @local_hmac_issuer
  def local_rsa_issuer, do: @local_rsa_issuer

  def local_issuer?(iss), do: iss in [@local_hmac_issuer, @local_rsa_issuer]

  def new, do: %{}

  def register(dispatcher, issuer, verifier_mod, verifier_arg) do
    Map.put(dispatcher, issuer, {verifier_mod, verifier_arg})
  end

  def verify(dispatcher, token) do
    with {:ok, [_header_b64, payload_b64, _sig_b64]} <- split(token),
         {:ok, payload} <- decode_payload(payload_b64),
         {:ok, iss} <- fetch_iss(payload),
         {:ok, {verifier_mod, verifier_arg}} <- lookup(dispatcher, iss) do
      verifier_mod.verify(verifier_arg, token)
    end
  end

  defp split(token) do
    case String.split(token, ".") do
      [_, _, _] = parts -> {:ok, parts}
      _ -> {:error, :malformed_token}
    end
  end

  defp decode_payload(payload_b64) do
    with {:ok, raw} <- Base.url_decode64(payload_b64, padding: false),
         {:ok, payload} <- Jason.decode(raw) do
      {:ok, payload}
    else
      _ -> {:error, :malformed_token}
    end
  end

  defp fetch_iss(%{"iss" => iss}) when is_binary(iss), do: {:ok, iss}
  defp fetch_iss(_), do: {:error, :unknown_issuer}

  defp lookup(dispatcher, iss) do
    case Map.fetch(dispatcher, iss) do
      {:ok, entry} -> {:ok, entry}
      :error -> {:error, :unknown_issuer}
    end
  end

  @doc false
  def claims_from_map(%{"sub" => sub, "iss" => iss} = payload) when is_binary(sub) and is_binary(iss) do
    azp = Map.get(payload, "azp", "")
    {:ok, %Claims{sub: sub, iss: iss, azp: azp}}
  end

  def claims_from_map(_), do: {:error, :missing_claims}
end
